[fix] bind:this works during onMount callback in manually-created component

rmunn opened the pull request:
If a component was instantiated with new Component(), its onMount callback was being called before the bind:this bindings were set up, so that the binding variable was undefined. Now the bind:this bindings are properly bound before the onMount callbacks are called.

Fixes #6760.
Fixes #6919.

### Before submitting the PR, please make sure you do the following
- [X] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [X] Prefix your PR title with `[feat]`, `[fix]`, `[chore]`, or `[docs]`.
- [X] This message body should clearly illustrate what problems it solves.
- [X] Ideally, include a test that fails without this PR but passes with it.

### Tests
-  [X] Run the tests with `npm test` and lint the project with `npm run lint`

Note that while this change makes the regression test I added (reproducing #6760 and #6919) pass, it makes the following tests fail:

* `await-catch-no-expression`
* `await-then-catch-multiple`
* `await-then-no-expression`

Somehow these tests are dependent on the current behavior of `flush()` that is causing #6760 and #6919. I have not yet found a way to fix those two issues and still have those three `await`-related tests pass, so I would need a little help figuring that out.

rmunn commented:
BTW, the current behavior of `flush` that I mentioned in the PR description is that if `flush()` is called while it is already running, it immediately returns without doing anything. The behavior prior to #4356 being merged was that `flush()` was reentrant. This PR essentially reverts #4356 and makes `flush()` reentrant again, but using a queue for `dirty_components` instead of an array so that `dirty_components.shift()` can be amortized O(1) instead of O(N). (The inefficiency of calling `shift()` N times in a while loop was the motivation for #4356, so I tried to maintain that efficiency.)

rmunn commented:
I can fix the three `await` tests with the following patch, but then the `onMount-get-current-component` test fails:

```js
diff --git a/src/runtime/internal/scheduler.ts b/src/runtime/internal/scheduler.ts
index cc93a6af5..fd69256a7 100644
--- a/src/runtime/internal/scheduler.ts
+++ b/src/runtime/internal/scheduler.ts
@@ -1,5 +1,6 @@
 import { run_all, Queue } from './utils';
 import { set_current_component } from './lifecycle';
+import { get_current_component } from '.';
 
 export const dirty_components = new Queue<any>();
 export const intros = { enabled: false };
@@ -37,12 +38,16 @@ export function flush() {
        do {
                // first, call beforeUpdate functions
                // and update components
+               let current_component = null;
+               try {
+                       current_component = get_current_component()
+               } catch {}
                while (dirty_components.length) {
                        const component = dirty_components.shift();
                        set_current_component(component);
                        update(component.$$);
                }
-               set_current_component(null);
+               set_current_component(current_component);
 
                dirty_components.length = 0;
 
```

The behavior of resetting the current component to `null` seems to be a deliberate feature of https://github.com/sveltejs/svelte/pull/4909, which means that there seems to be a conflict between three features:


* #6564 works if `flush` is not reentrant, *or* if `flush` is reentrant and **does not** clear the current component
* #4909 works if `flush` is not reentrant, *or* if `flush` is reentrant and **does** clear the current component
* This PR works if `flush` is reentrant

So if we make `flush` reentrant, either #6564 or #4909 will fail. If we do not make `flush` reentrant, this PR will fail and #6760 and #6919 will not be fixed.

Paging @tanhauhau who can help me understand why #6564 fails when `flush` clears the current component. Because it sounds like #4909 absolutely relies on that behavior, so I have to find a way to make #6564 work without losing the `set_current_component(null)` behavior.

rmunn commented:
Found a way to make all the unit tests pass, by moving the "save and restore current component" outside the do ... while loop so that it only runs once. This seems to fulfill all the requirements of the previous work on the `flush()` function:

* Calling `get_current_component()` during `onMount()` still throws an exception (#4909)
* `flush()` is now reentrant again, undoing the `if (flushing) return` change from #4356
    * See below for why this is OK
* Tests from #6564 still pass

Reasons why undoing the `if (flushing) return` change from #4356 is okay:

* The motivation for #4356 was efficiency: `shift()` is O(N) on an array, and `flush()` calls it O(N) times
* The `if (flushing) return` change in PR 4356 was required because the change from a `while` loop to a `for` loop in that PR caused an infinite loop when `flush()` was called reentrantly (see PR 4356 discussion for details)
* Making `flush()` non-reentrant was not the main motivation for PR 4356
* With this PR, `dirty_components` is now a queue instead of an array, allowing `shift()` to work in amortized O(1) time
* This PR brings back the `while` loop that existed before PR 4356, so that it's safe for `flush()` to be reentrant again
* A reentrant `flush()` is necessary for `bind:this` to function properly during `onMount()`

This should be ready for someone to review and merge now.

rmunn commented:
While I wait for a Svelte team member to have time to review this, a couple words about the Queue implementation. I chose to go with a deliberately minimal implementation, with only the bare minimum features needed to fulfill the needs of the `dirty_components` queue. So I only implemented `push`, `shift`, and `length`. I probably didn't need to implement the ability to set the length of the queue to 0 (clearing it), but I did need to implement a getter on length and it seemed silly not to make it a getter/setter. But there's no iterators and no index lookup, because they're not currently needed in the `flush()` implementation. I figure if they're needed in the future, they can be added later. You Ain't Gonna Need It and all that.

benmccann commented on src/runtime/internal/scheduler.ts:
```diff
@@ -1,7 +1,7 @@
+import { get_current_component, set_current_component } from './lifecycle';
 
-export const dirty_components = [];
+export const dirty_components = new Queue<any>();
```
is there a component type that could be used here rather than `any`?

rmunn replied:
Good question. I'd like to know the answer myself. :-) The type specified here would need to be the same type that is used in `get_current_component` and `set_current_component` in `src/runtime/internal/lifecycle.ts`. Both of them are currently untyped, so Typescript assigns them a type of `any`. However, I tried doing `import type { SvelteComponent } from './Component'` and then making this a `Queue<SvelteComponent>`, and TypeScript was apparently happy with that.

So it looks like `SvelteComponent` would be the right type to put here. But there's also `SvelteComponentDev` and `SvelteComponentTyped` listed in `src/runtime/index.ts`, and I'm not 100% sure what the differences are. There's also a comment at the definition of SvelteComponentTyped pointing to https://github.com/microsoft/TypeScript/issues/41770 as the reason for its existence, so perhaps @dummdidumm would know more about this.

rmunn replied:
Update: both SvelteComponentDev and SvelteComponentTyped are subclasses of SvelteComponent, so it looks like SvelteComponent might indeed be the right type to put in here. I'm still hesitant to make that change because `dirty_components` was previously untyped, and I don't know if making its type stricter than before might break anyone's existing Svelte code.

bluwy commented on src/runtime/internal/scheduler.ts:
```diff
@@ -31,17 +31,21 @@ export function add_flush_callback(fn) {
 		// and update components
-		for (let i = 0; i < dirty_components.length; i += 1) {
-			const component = dirty_components[i];
+		while (dirty_components.length) {
+			const component = dirty_components.shift();
```
Is the queue change required for the issues fixed? Otherwise I think we can separate the PR as it's a crucial decision to make. It may be hard to justify the performance gain for the extra bundle size.

dummdidumm replied:
Is there any way to make this work without the queue? Like tracking the current index outside of the flush function and resetting it when flush is done.

rmunn replied:
Thanks for the idea, @dummdidumm. That worked, so I've implemented it (and removed the Queue implementation) in https://github.com/sveltejs/svelte/pull/6920/commits/83e9cc5d5d405c9cbd20176c0bc636bb782178a6. Note that I put the index increment on its own line instead of doing `dirty_components[flushidx++]`. I've never liked the `array[idx++]` and `array[++x]` idioms; I view them as developer traps, mistakes waiting to happen. Any time it's possible to be more explicit about whether the increment is happening before or after the array deference, I prefer to be explicit.

But if you (plural you, the core Svelte team) want this to be written as `dirty_components[flushidx++]`, I'm willing to make that change.

bluwy commented on src/runtime/internal/scheduler.ts:
```diff
@@ -31,17 +31,21 @@ export function add_flush_callback(fn) {
+	try {
+		current_component = get_current_component();
+	} catch {
+		// no current component, so leave it as null
```
I think we can explain why it's safe to leave it empty, instead of what

dummdidumm replied:
How often will this throw? Catching errors is expensive.

rmunn replied:
It throws quite often — actually in the majority of cases, I think — so I added a new `maybe_get_current_component` function in https://github.com/sveltejs/svelte/pull/6920/commits/47f90d0a1f30a3e973313bbf8c7544e11ed42993 to avoid throwing and catching useless errors when all we want (in the `flush()` function, at least) is to save and restore the value, without caring what it was.

rmunn replied:
Note that I could also have just had `flush()` close over the `current_component` value from `lifecycle.ts`. It's exported, after all. I wasn't sure if that could produce a stale-closures issue, but I just tried it and it passed all the unit tests, so it's probably never going to be stale. I'll push one more commit getting rid of the `maybe_get_current_component` function.

rmunn replied:
Just pushed https://github.com/sveltejs/svelte/pull/6920/commits/407213cd551f271db30b848b724cbe6430e385b1 which doesn't add a new function, and instead has `flush()` access `current_component` directly rather than calling `get_current_component()` and catching errors.

rmunn commented:
One unit test timed out after 15 minutes, apparently because the test runner got stuck. I don't have the access rights to re-run the tests. Could someone push the "Re-run all checks" button for me? Or I could push a meaningless rebase to force the tests to re-run, but that seems a little silly.

benmccann commented:
We basically just ignore it if it's only a single job failing since the tests are known to be flaky

rmunn commented:
> We basically just ignore it if it's only a single job failing since the tests are known to be flaky

Good to know. Turned out I'd accidentally missed removing one blank line when I removed the code block above it, so I did end up pushing a commit making no code changes, which ended up re-running all the unit tests. And then one of the Mac test runners had a browser that died, which cancelled seven different jobs. So it ended up being a bit of a moot point, because I did have to do a `commit --amend` and a force-push in order to get the tests looking greenish again. (This time two Windows tests timed out, on different Node versions than the one that timed out earlier).

Now that I've simplified the code by removing the Queue implementation (thanks to the good suggestion @dummdidumm made), I think most of this PR will be easy to approve. There are two points remaining that I'd like someone to weigh in on:

* How I should handle saving and restoring the current component inside `flush()`. Three possible ways of doing it:
    * (Current state of the code): access `current_component` directly inside `flush()`, without using `get_current_component()`
    * Call `get_current_component()` and catch the error if it's undefined/null
    * Create a new `maybe_get_current_component()` function that doesn't throw, and use that
* Is having `flushidx++` on its own line a coding style that the Svelte team likes? Or would you prefer that I move it inside the array lookup, so that it becomes `const component = dirty_components[flushidx++]` in one line instead of two?

Pros and cons of the ways to save and restore the current component:
* Direct access. Pro: Simple, doesn't create new code. Con: if `get_current_component()` grows more logic in the future, `flush()` won't benefit from it.
* Throw and catch. Pro: doesn't create new code in `lifecycle.ts`. Con: expensive in performance: will be catching and ignoring errors quite frequently.
* New `maybe_get_current_component()` function. Pro: if `get_current_component` grows more logic in the future, it will be clear that `maybe_get_current_component` should do the same. Con: adds a new function in `lifecycle.ts`, and it will probably be unclear in the future why these two functions differ.

dummdidumm commented:
Thanks for your work on this, and especially for laying out your thought process this clearly.

My opinion on your questions:
- I agree that it's more readable to do it on its own line. On the other hand, it saves a few bites if you do it inline. I'm a little indifferent but lean more to the "more readable" side
- current component: Due to the bundle size concerns just mentioned I'd go with accessing the variable directly, maybe adding a little explanation why `get_current_component` isn't used (might throw frequently in here which is a performance hit)

In general, while we're at it, this specific method could use some more comments in my opinion. I feel like you looked at this function from all angles by now and can probably confidently add some comments explaining why things are done the way, for example that flush can be reentrant (and why), why the index needs to be tracked outside of the functions, etc.

rmunn commented:
@dummdidumm wrote:
> In general, while we're at it, this specific method could use some more comments in my opinion. I feel like you looked at this function from all angles by now and can probably confidently add some comments explaining why things are done the way, for example that flush can be reentrant (and why), why the index needs to be tracked outside of the functions, etc.

Done in commit fd7586c44. Let me know if that's a bit too much and I'll trim the comments down a little.

rmunn commented:
Note https://github.com/sveltejs/svelte/issues/6016 - I need to write another unit test to see if this PR would fix #6016 or whether that bug needs a different fix. Because https://github.com/sveltejs/svelte/pull/6858, intended to fix #6016, might conflict with this PR. More research needed.

rmunn commented:
Pulled #6858's unit tests into my branch and ran them, and my changes do *not* fix the bugs that #6858 fixes, where `beforeUpdate` gets called twice. I then pulled #6858's changes into my branch and abandoned my branch's changes to `flush()`, keeping my branch's unit tests for `bind:this` being null, and #6858 does *not* fix the bug that this PR fixes.

So both this PR and #6858 are needed as they fix different issues. Whichever one is merged first will cause a merge conflict in the other one, but they should be simple enough to resolve.

dummdidumm reviewed (approved):
I like the comment! By now (and through the comment) I think I understood enough of this part to give my approval on this PR 👍 If this one's merged, you could then look at the other PR you mentioned since you are already so familiar with that part, if you like.

rmunn commented:
Another issue that I should investigate is #6921. I expect it wouldn't be fixed by this PR, but it's quite possible that #6858 would fix it.

bluwy commented on src/runtime/internal/scheduler.ts:
```diff
@@ -31,23 +31,42 @@ export function add_flush_callback(fn) {
+//    reverse order: children before parents.
+// Since callbacks might update component values, which could trigger another
+// call to flush(), the following steps guard against this:
+// 1. During beforeUpdate, any updated components will be added to the
+// dirty_components array and will cause a reentrant call to flush(). Because
+// the flush index is kept outside the function, the reentrant call will pick
+// up where the earlier call left off and go through all dirty components. The
+// current_component value is saved and restored so that the reentrant call will
+// not interfere with the "parent" flush() call.
```
```suggestion
// 1. During beforeUpdate, any updated components will be added to the
//    dirty_components array and will cause a reentrant call to flush(). Because
//    the flush index is kept outside the function, the reentrant call will pick
//    up where the earlier call left off and go through all dirty components. The
//    current_component value is saved and restored so that the reentrant call will
//    not interfere with the "parent" flush() call.
```
I think it's missing an indent here.
